src/Workspaces/Core/Portable/Execution/Serialization/Checksum.cs

```diff
@@ -73,8 +75,8 @@ public override int GetHashCode()
                     hash = Hash.Combine((int)_checkSum[i], hash);
                 }
 
-                _lazyHash = hash;
-            }
+                _lazyHash = new StrongBox<int>(hash);
+            };
```

Comment:
> hmmm.. I thought reference assignment is atomic in .NET.

The assignment has atomic.  But htat doens't mean that all hte writes that occurred before it will be visible to someone reading it. 

>  there is so many bugs in compiler since compiler does what I am doing all over the places.

That should not be hte case.  The compiler is very careful to use appropriate constructs in many/most places to make this safe . If you see any places like this, please point them out so we can file bugs and fix them.